executor/parallel_NLJ.go

```diff
@@ -0,0 +1,318 @@
+	e.innerSelected = make([][]bool, e.concurrency)
+	e.innerIter = make([]chunk.Iterator, e.concurrency)
+	e.outerRow = make([]*chunk.Row, e.concurrency)
+	e.hasMatch = make([]bool, e.concurrency)
+	e.hasNull = make([]bool, e.concurrency)
+	for i := 0; i < e.concurrency; i++ {
+		e.innerChunk[i] = newFirstChunk(e.innerExecs[i])
+		e.innerList[i] = chunk.NewList(retTypes(e.innerExecs[i]), e.initCap, e.maxChunkSize)
+		e.innerList[i].GetMemTracker().SetLabel(innerListLabel)
+		e.innerList[i].GetMemTracker().AttachTo(e.memTracker)
+	}
+
+	e.freeChkCh = make(chan *chunk.Chunk, e.concurrency)
+	e.resultChkCh = make(chan result, e.concurrency+1) // innerWorkers + outerWorker
+	e.outerRowCh = make(chan outerRow)
+	e.exit = make(chan struct{})
+	for i := 0; i < e.concurrency; i++ {
+		e.freeChkCh <- newFirstChunk(e)
+	}
+	return nil
+}
+
+// Next implements the Executor interface.
+func (e *ParallelNestedLoopApplyExec) Next(ctx context.Context, req *chunk.Chunk) (err error) {
+	if !e.started {
+		e.wg.Add(1)
+		go e.outerWorker(ctx)
+		for i := 0; i < e.concurrency; i++ {
+			e.wg.Add(1)
+			atomic.AddInt32(&e.numWorkers, 1)
+			go e.innerWorker(ctx, i)
+		}
+		e.started = true
+	}
+	result := <-e.resultChkCh
+	if result.err != nil {
+		return result.err
+	}
+	req.SwapColumns(result.chk)
+	e.freeChkCh <- result.chk
+	return nil
+}
+
+// Close implements the Executor interface.
+func (e *ParallelNestedLoopApplyExec) Close() error {
+	e.memTracker = nil
+	err := e.outerExec.Close()
+	if e.started {
+		close(e.exit)
+		e.started = false
+		e.wg.Wait()
+	}
+	return err
+}
+
+func (e *ParallelNestedLoopApplyExec) outerWorker(ctx context.Context) {
+	defer e.handleWorkerPanic(ctx)
+	var selected []bool
+	var err error
+	for {
+		chk := newFirstChunk(e.outerExec)
+		if err := Next(ctx, e.outerExec, chk); err != nil {
+			e.putResult(nil, err)
+			return
+		}
+		if chk.NumRows() == 0 {
+			close(e.outerRowCh)
+			return
+		}
+		e.outerList.Add(chk)
+		outerIter := chunk.NewIterator4Chunk(chk)
+		selected, err = expression.VectorizedFilter(e.ctx, e.outerFilter, outerIter, selected)
+		if err != nil {
+			e.putResult(nil, err)
+			return
+		}
+		for i := 0; i < chk.NumRows(); i++ {
+			row := chk.GetRow(i)
+			select {
+			case e.outerRowCh <- outerRow{&row, selected[i]}:
+			case <-e.exit:
+				return
+			}
+		}
+	}
+}
+
+func (e *ParallelNestedLoopApplyExec) innerWorker(ctx context.Context, id int) {
+	defer e.handleWorkerPanic(ctx)
+	for {
+		var chk *chunk.Chunk
+		select {
+		case chk = <-e.freeChkCh:
+		case <-e.exit:
+			return
+		}
+		err := e.fillInnerChunk(ctx, id, chk)
+		if err == nil && chk.NumRows() == 0 { // no more data, this goroutine can exit
+			for {
+				numWorkers := atomic.LoadInt32(&e.numWorkers) // decrease the numWorkers counter
+				if atomic.CompareAndSwapInt32(&e.numWorkers, numWorkers, numWorkers-1) {
+					if numWorkers == 1 { // it's the last one, deliver an empty chunk to notify the upper executor
```

Comment:
It is not safe to do this. We can do this after `e.wg.done`.

Reply:
Could you please explain this more detailedly? How to do this after `e.wg.done`. @fzhedu

Reply:
use another goroutine or put it in the outerworker?

Reply:
It is updated. I introduce a new goroutine (notifyWorker) to wait for all inner/outer-workers finishing and then put an empty chunk to notify the upper executor there is no more data.

PTAL @fzhedu